### PR TITLE
dialects: scf.if make else region optional

### LINKED ISSUE #372: Make else_region optional in scf.if

### PR DESCRIPTION
**Pull request by AntonLydike**
Fixes #372

### CODE REVIEW

File: xdsl/dialects/scf.py [hunk 1]
```diff
@@ -21,9 +21,16 @@ class If(Operation):
...
-            false_region: Region | List[Block] | List[Operation]) -> If:
+    def get(
+        cond: SSAValue | Operation,
+        return_types: List[Attribute],
+        true_region: Region | List[Block] | List[Operation],
+        false_region: Region | List[Block] | List[Operation] | None = None
+    ) -> If:
+
+        if false_region is None:
+            false_region = Region.from_operation_list([Yield.get()])
```
Comment (PapyChacal) [inline xdsl/dialects/scf.py:32]: MLIR models that by a Region with no block, can't we do the same?

**Review by georgebisbas — CHANGES_REQUESTED** [review]
Should we add a test that showcases an scf.if with no else region?

**Comment by math-fehr** [comment]
Oh, I indeed made a mistake when I wrote this issue :/
MLIR seems to handle empty `else` with a region with no blocks. This is weird, but I think we should do the same, since we wouldn't be compatible with them otherwise :(

File: xdsl/dialects/scf.py (hunk 1 above)
Comment (AntonLydike) [reply xdsl/dialects/scf.py]: Yep! Done!

**Comment by AntonLydike** [comment]
> Oh, I indeed made a mistake when I wrote this issue :/

I don't think you did, it was just my interpretation of it that was wrong^^

**Comment by webmiche** [comment]
I can look into the structural equivalence thing. Personally, I would just choose a different test case for the change you are doing here, one that does not rely on structural equivalence, but rather checks that there is an empty block directly.

**Review by superlopuh — APPROVED** [review]

**Comment by webmiche** [comment]
Nice, this is MLIR screwing with us big time: AFAIR, function declarations have regions with an empty block (printed `({})`), where as this seems to be an empty region (printed `({})`). Not sure how we want to handle this. I guess empty regions parsing would have to be operation specific, which is just ridiculous, IMO.

Did we verify that this really is an empty region, in one with no Blocks?

File: tests/dialects/test_scf.py [hunk 2]
```diff
@@ -52,3 +58,23 @@ def test_parallel():
...
     assert p.body is body
+
+
+def test_empty_else():
+    m = ModuleOp.from_region_or_ops([
+        t := Constant.from_int_and_width(1, 1),
+        If.get(t, [], [
+            Yield.get(),
+        ]),
+    ])
```
Comment (superlopuh) [inline tests/dialects/test_scf.py:69]:
```suggestion
    ])
    assert len(cast(If, m.ops[1]).else_.blocks) == 0
```

**Comment by AntonLydike** [comment]
I changed the test so it doesn't test printing/parsing. This will not *fix* the brokeness here, just hide it for now.

**Comment by AntonLydike** [comment]
@georgebisbas 
> Should we add a test that showcases an scf.if with no else region?

I added a test for it, does it look good?

**Review by PapyChacal — APPROVED** [review]
You also could make the non-returning `scf.if` implicitly add the empty `scf.yield` at the end of either regions if misssing and required. But that's no blocker, looks good already to me!

**Review by math-fehr — APPROVED** [review]
:+1:

**Comment by AntonLydike** [comment]
@math-fehr @PapyChacal @georgebisbas feel free to merge when all is green, I'm away for the next couple of days!

**Review by meshtag — APPROVED** [review]
Not sure about this, but do we add a filecheck test as well?

**Review by webmiche — APPROVED** [review]

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**